tornado/test/httpserver_test.py

```diff
@@ -834,37 +834,53 @@ def setUp(self):
         app = Application([("/hello", HelloWorldRequestHandler)])
         self.server = HTTPServer(app)
-        self.server.add_socket(sock)
-        self.stream = IOStream(socket.socket(socket.AF_UNIX))
-        self.io_loop.run_sync(lambda: self.stream.connect(self.sockfile))
+        if sys.platform.startswith("linux"):
+            self.sockabstract = "\0" + os.path.basename(self.tmpdir)
+            self.server.add_socket(netutil.bind_unix_socket(self.sockabstract))
+        self.server.add_socket(netutil.bind_unix_socket(self.sockfile))
 
     def tearDown(self):
-        self.stream.close()
         self.io_loop.run_sync(self.server.close_all_connections)
         self.server.stop()
         shutil.rmtree(self.tmpdir)
         super().tearDown()
 
     @gen_test
     def test_unix_socket(self):
-        self.stream.write(b"GET /hello HTTP/1.0\r\n\r\n")
-        response = yield self.stream.read_until(b"\r\n")
-        self.assertEqual(response, b"HTTP/1.1 200 OK\r\n")
-        header_data = yield self.stream.read_until(b"\r\n\r\n")
-        headers = HTTPHeaders.parse(header_data.decode("latin1"))
-        body = yield self.stream.read_bytes(int(headers["Content-Length"]))
-        self.assertEqual(body, b"Hello world")
+        with closing(IOStream(socket.socket(socket.AF_UNIX))) as stream:
+            stream.connect(self.sockfile)
+            stream.write(b"GET /hello HTTP/1.0\r\n\r\n")
+            response = yield stream.read_until(b"\r\n")
+            self.assertEqual(response, b"HTTP/1.1 200 OK\r\n")
+            header_data = yield stream.read_until(b"\r\n\r\n")
+            headers = HTTPHeaders.parse(header_data.decode("latin1"))
+            body = yield stream.read_bytes(int(headers["Content-Length"]))
+            self.assertEqual(body, b"Hello world")
+
+    @unittest.skipUnless(sys.platform.startswith("linux"), "requires Linux")
+    @gen_test
+    def test_unix_socket_abstract(self):
+        with closing(IOStream(socket.socket(socket.AF_UNIX))) as stream:
+            stream.connect(self.sockabstract)
+            stream.write(b"GET /hello HTTP/1.0\r\n\r\n")
+            response = yield stream.read_until(b"\r\n")
+            self.assertEqual(response, b"HTTP/1.1 200 OK\r\n")
+            header_data = yield stream.read_until(b"\r\n\r\n")
+            headers = HTTPHeaders.parse(header_data.decode("latin1"))
+            body = yield stream.read_bytes(int(headers["Content-Length"]))
+            self.assertEqual(body, b"Hello world")
 
     @gen_test
     def test_unix_socket_bad_request(self):
         # Unix sockets don't have remote addresses so they just return an
         # empty string.
         with ExpectLog(gen_log, "Malformed HTTP message from", level=logging.INFO):
-            self.stream.write(b"garbage\r\n\r\n")
-            response = yield self.stream.read_until_close()
+            with closing(IOStream(socket.socket(socket.AF_UNIX))) as stream:
+                stream.connect(self.sockfile)
```

Comment:
This test is about introspecting socket addresses, so it feels like something that should be tested for abstract sockets too. Instead of binding both abstract and filesystem sockets in setUp (half of which is going to be wasted work in any given test), and only testing the abstract socket with certain tests that must be duplicated, I think it would be better to use subclasses: factor the socket binding logic out of setUp into an overridable method, and then make an AbstractUnixSocketTest that subclasses this one.

Reply:
Refactored unit-tests as suggested in https://github.com/tornadoweb/tornado/pull/3405/commits/be9e153b36fe3161394ebdf58c6d4aa426836821.

Reply:
Sorry, I'm not getting on very well with the pull-request checks…

Any hints as to how you'd like to make [mypy](https://mypy.readthedocs.io/) happy about the refactored test case. I didn't want to make the `UnixSocketTest` base class an [AsyncTestCase](https://www.tornadoweb.org/en/stable/testing.html#tornado.testing.AsyncTestCase) because I don't want the test functions to be executed outside of one of the, `UnixSocketTestAbstract` or `UnixSocketTestFile` concrete implementations; but a failure to do so makes [mypy](https://mypy.readthedocs.io/) sad.

Reply:
Hmm, I thought I had a clever solution for this but it looks like the other places where I've done this are just hacks with a lot of `typing.Any` and `type:ignore` annotations. (Other discussions of the issue at https://github.com/abseil/abseil-py/issues/166 and https://github.com/python/cpython/issues/61721)

In this case I think I'd make `UnixSocketTest` a subclass of `AsyncTestCase` to make type checking happy, and then skip the base class at runtime: In `UnixSocketTest.setUp()`, add `if type(self) == UnixSocketTest: raise unittest.SkipTest("abstract base class")`. That might have the fortunate side effect of making pytest happier with this pattern too (#3390)

Reply:
Implemented as suggested in [6e14f25](https://github.com/tornadoweb/tornado/compare/1ddd3d5f0f0a4740ca58ed4702bf166f683069b4..6e14f25fcd95d505e2252a5b42ee7aeb4e4315a1).